### PR TITLE
Migrate `emit-to-stdout` `run-make` test to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Another test which will fail CI. Most of everything here works, with the only exception being the cases where the test redirects stdout to a tty: `1>/dev/ptmx`.

`run_make_support` doesn't support this yet, so I am posting this PR to open a discussion.

### CODE REVIEW

**Comment by jieyouxu** [comment]
Note to self: look at the pseudoterminal stuff

**Comment by ChrisDenton** [comment]
Note that the test does not necessarily need `/dev/ptmx` specifically. It's just testing that printing an object to a terminal fails. The problem being that the test framework redirects stdout. Maybe there's another way to print to an actual terminal, even if we have to spawn one?

**Comment by ChrisDenton** [comment]
Btw, on windows this can be done via [`CommandExt::creation_flags`](https://doc.rust-lang.org/std/os/windows/process/trait.CommandExt.html#tymethod.creation_flags) with `CREATE_NEW_CONSOLE` (aka `0x10`).

**Comment by jieyouxu** [comment]
Progress update: none, since I haven't gotten to this one yet, lol

File: tests/run-make/emit-to-stdout/rmake.rs
Comment (jieyouxu) [inline tests/run-make/emit-to-stdout/rmake.rs:1]: Suggestion: for this test, I find the complex conditional control flow in `test_emit` quite hard to follow. Can we structure it more similarly to the original Makefile?

```rs
fn main() {
    test_asm();
}

fn test_asm() {
    run_in_tmpdir(|| {
        rustc().emit(format!("asm={}", whatever)).input(test.rs).run();
		rustc().emit("asm=-").input(test.rs).run();
    });
}
```

I know this doesn't fix the other problems.

**Review by jieyouxu — CHANGES_REQUESTED** [review]

File: tests/run-make/emit-to-stdout/rmake.rs
Comment (Oneirical) [reply tests/run-make/emit-to-stdout/rmake.rs:1]: I feel reluctant to divide the test into the original control flow, because it implies a significant amount of repetition. A little typo sneaking into one of the cases would be harder to spot.

I have added explanatory comments to try to make `test_emit` easier to follow.

If you disagree with me, restate that you think it is better to split it like the original Makefile, and I will comply.

Comment (Oneirical) [reply tests/run-make/emit-to-stdout/rmake.rs:1]: Very well, it is done.

I've been looking into some ways to fix the `dev/ptmx` problem. [This StackOverflow thread seems to be of interest](https://stackoverflow.com/questions/60367318/how-do-you-write-to-a-pty-master-rust). I'd like to hear your thoughts on it.

@rustbot review

Comment (jieyouxu) [reply tests/run-make/emit-to-stdout/rmake.rs:1]: Yeah, I think we might be able to create our own pty. I'll try to see if I can something work around tomorrow. I've been busy trying to debug why our msvc CI jobs are failing so frequently lol.

**Comment by jieyouxu** [comment]
Review notes:

- Related issue: https://github.com/rust-lang/rust/issues/119658
- Related PR: https://github.com/rust-lang/rust/pull/119664

File: tests/run-make/emit-to-stdout/rmake.rs
Comment (jieyouxu) [inline tests/run-make/emit-to-stdout/rmake.rs:1]: Suggestion: rename this test to `rustc-emit-to-stdout` or something, since apparently there's a `emit-shared-files` test which is exercising rustdoc only (where I suggested to rename that test to `rustdoc-emit-shared-files` as well).

**Comment by ChrisDenton** [comment]
Here's my (very quick) implementation that might work on Linux. I've not tested it though. df8d4ba8b547ee700763754b877009b94f5c147c

It works on Windows but note the "fixme".

**Comment by jieyouxu** [comment]
Closing this in favor of #131355 where we consolidated all of the efforts for `emit-to-stdout`.
